### PR TITLE
Should use 0.0.0.0 in agent mbus url for director

### PR DESCRIPTION
**Pull request by mattcui**
For addressing tracker [#133906403](https://www.pivotaltracker.com/story/show/133906403), 

@maximilien  Please review this PR after [PR 216](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/pull/216) is merged, thanks.

### CODE REVIEW

File: src/bosh-softlayer-cpi/action/create_vm.go [hunk 1]
```diff
@@ -77,10 +77,10 @@ func (a CreateVMAction) Run(agentID string, stemcellCID StemcellCID, cloudProps
 func (a CreateVMAction) updateCloudProperties(cloudProps *VMCloudProperties) {
 	a.vmCloudProperties = cloudProps
 
-	if len(cloudProps.BoshIp) == 0 || cloudProps.Baremetal {
-		a.vmCloudProperties.VmNamePrefix = cloudProps.VmNamePrefix
+	if cloudProps.DeployedByBoshCLI {
```
Comment (maximilien) [inline src/bosh-softlayer-cpi/action/create_vm.go]: This looks similar to previous commit. Might need to rebase

File: src/bosh-softlayer-cpi/bin/test_unit [hunk 2]
```diff
@@ -19,13 +23,9 @@
   echo -e "\n Formatting packages..."
   go fmt ./...
 
-  echo -e "\n Make the build..."
-  go build -o out/cpi bosh-softlayer-cpi/main
-
-  echo -e "\n Integration Testing packages:"
-  ginkgo -r -p --noisyPendings --skipPackage=integration
+  echo -e "\n Unit Testing packages:"
```
Comment (maximilien) [inline src/bosh-softlayer-cpi/bin/test_unit]: I think this is already changed in previous PR. Please rebase

**Comment by mattcui** [comment]
@maximilien Rebased, please review, thanks.

**Comment by mattcui** [comment]
@maximilien corrected some typos, it's urgent for review and merge. Thanks.

**Comment by mattcui** [comment]
Hi @maximilien, I added some UT code to verify if mbus in agentOptions is as expected. Since agentOptions is a private variable, I have to add a new public function to retrieve the value of agentOptions. Please review. Thanks.

**Comment by maximilien** [comment]
OK, see my comment about `VcapPassword` now being visible and needing to be obfuscated? See some of my suggestions in the comment.

**Comment by mattcui** [comment]
Hi @maximilien, it's strange that I can't find your comment about VcapPassword, could you please post it as a general comment here again? Thanks.

**Comment by maximilien** [comment]
My main concern here is that now the VcapPassword which is pretty serious information is now accessible in all other packages and could potentially be made public in future PRs or when logging etc.

I wonder if we could avoid making the entire AgentOptions public? or does VcapPassword need to be public? or perhaps when returning return c.agentOptions here you simply set the VcapPassword to ***** in the copied structure?

**Comment by maximilien** [comment]
See ^^^

**Comment by maximilien** [comment]
OK what's the status of this. Depending PR is now merged.

File: src/bosh-softlayer-cpi/softlayer/pool/softlayer_pool_creator.go [hunk 3]
```diff
@@ -150,6 +150,8 @@ func (c *softLayerPoolCreator) createFromVMPool(agentID string, stemcell bslcste
 	return sl_vm_os, nil
 }
 
+func (c *softLayerPoolCreator) GetAgentOptions() AgentOptions { return c.agentOptions }
```
Comment (mattcui) [inline src/bosh-softlayer-cpi/softlayer/pool/softlayer_pool_creator.go]: @maximilien Do you think if it's feasible to change `VcapPassowrd` defined in `AgentOptions` to a private variable, then write a public function `GetVcapPassword` to retrieve the vcap password in `agent_options.go`? Thanks.

**Comment by mattcui** [comment]
@maximilien I changed vcap password to a private variable, and use public Get/Set functions to manipulate it. Please help review my changes, thanks.

File: src/bosh-softlayer-cpi/softlayer/common/agent_options.go [hunk 4]
```diff
@@ -42,6 +42,14 @@ func (o AgentOptions) Validate() error {
 	return nil
 }
 
+func (o *AgentOptions) GetVcapPassword() string {
+	return o.vcapPassword
+}
+
+func (o *AgentOptions) SetVcapPassword(vcapPassword string) {
```
Comment (maximilien) [inline src/bosh-softlayer-cpi/softlayer/common/agent_options.go]: @mattcui 

But this still makes the changing and getting of the password public? So nothing changed.

I think another way to phrase the question I am trying to ask is whether `VcapPassword` can be private and only used within the package? Or does it have to be public?

File: src/bosh-softlayer-cpi/softlayer/common/agent_options_test.go [hunk 5]
```diff
@@ -47,6 +47,21 @@ var _ = Describe("AgentOptions", func() {
 			Expect(err.Error()).To(ContainSubstring("Validating Blobstore configuration"))
 		})
 	})
+
+	Describe("Get/SetVcapPassword", func() {
+		Context("when vcapPassword is not set by SetvcapPassword()", func() {
```
Comment (maximilien) [inline src/bosh-softlayer-cpi/softlayer/common/agent_options_test.go]: Also fix this to be: `"when vcapPassword is not set by SetVcapPassword()"`

File: src/bosh-softlayer-cpi/softlayer/common/agent_options.go (hunk 4 above)
Comment (maximilien) [inline src/bosh-softlayer-cpi/softlayer/common/agent_options.go]: Let's discuss tonight and decide right then and there. Seems that this is taking way more time than you or I want.

Comment (mattcui) [inline src/bosh-softlayer-cpi/softlayer/common/agent_options.go]: @maximilien I reviewed the code, vcap password has to be used in other packages, like [softlayer_virtual_guest_creator.go](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/vm/softlayer_virtual_guest_creator.go#L146). But please note that the password is encrypted with SHA-512 (when it's set in cpi.json). Do you think if it's still a case? Thanks.

**Comment by mattcui** [comment]
@maximilien I modified the code and please review. Thanks.

**Comment by maximilien** [comment]
@mattcui ,

OK merging. But I really would like @zhanggbj and @gu-bin to also double check this in pipeline, making sure all is fine.

FYI that increasingly I would like you all to peer-review each other's code. I will request that, make sure you do so and confirm pipeline as well.

Best,

Max

**Pull request merged**